### PR TITLE
[client] reduce the number of allocations

### PR DESCRIPTION
**Pull request by truthbk**
This is an adaptation of #74 by @njhill by which we take the same principles and apply them to the new client architecture to reduce the overhead of allocations on the client, and consequently on the client application.

Still WIP - a couple of things went wrong after a rebase, so a few things still pending fixes.

### CODE REVIEW

**Comment by njhill** [comment]
@truthbk I started to review yesterday before you made further changes and had actually already written a comment suggesting to have a separate class for the sender `Runnable`s as I see you've now added! Curious why you added separate lists for the `StringBuilder`s/`CharBuffer`s though - why not just have a dedicated instance in each class (in instance fields)? Then you could move the `writeBuilderToSendBuffer` method to that class too, which would allow for better reuse of the `CharBuffer`s as in the original PR.

I'd also suggest to move other things into this class so that each worker has its own instance since they are not threadsafe:
- The `CharsetEncoder` (`utf8Encoder`)
- The `NumberFormat` instances

Anyhow let me know when it is ready and I'll take a look again :)

**Comment by truthbk** [comment]
Thank you for taking a look @njhill! The goal is to do all of that too! I'm fixing things one step at a time. Don't spend too much time reviewing this until the WIP tag is removed from the PR :)

In any case, thank you for the feedback, I had the exact same thoughts in mind to avoid code duplication and ensure a clean code structure. I still hadn't gone into verifying the thready safety of `CharsetEncoder` and `NumberFormat`, so thank you for that. 

I'll let you know when this is ready!

**Comment by truthbk** [comment]
@njhill open for review if you want to chime in! :)

File: src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java [hunk 1]
```diff
@@ -70,49 +72,51 @@
     };
 
     /**
-     * Because NumberFormat is not thread-safe we cannot share instances across threads. Use a ThreadLocal to
-     * create one pre thread as this seems to offer a significant performance improvement over creating one per-thread:
-     * http://stackoverflow.com/a/1285297/2648
-     * https://github.com/indeedeng/java-dogstatsd-client/issues/4
+     * The NumberFormat instances are not threadsafe but are only ever called from
+     * the sender thread.
```
Comment (njhill) [inline src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java]: This comment now obsolete

**Comment by truthbk** [comment]
Gotta fix a few things after the rebase to telemetry - don't panic! 😉

File: src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java (hunk 1 above)
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java]: Updated it.

File: src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java [hunk 2]
```diff
@@ -1,29 +1,114 @@
...
 
+
 public class StatsDNonBlockingProcessor extends StatsDProcessor {
 
-    private final Queue<String> messages;
+    private final Queue<Message> messages;
     private final int qcapacity;
     private final AtomicInteger qsize;  // qSize will not reflect actual size, but a close estimate.
 
+    private class ProcessingTask extends StatsDProcessor.ProcessingTask {
```
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java]: They're pretty close but the logic dealing with the queues blocking vs non-blocking is a little different. I might clean that up, but would rather do it in a separate PR.

File: src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java [hunk 3]
```diff
@@ -1,29 +1,114 @@
...
+
+                try {
+                    if (empty) {
+                        Thread.sleep(WAIT_SLEEP_MS);
+                        continue;
+                    }
+
+                    if (Thread.interrupted()) {
+                        return;
+                    }
```
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java]: Agreed, just remnants from old code, we can clean that up.

File: src/main/java/com/timgroup/statsd/StatsDProcessor.java [hunk 4]
```diff
@@ -28,6 +40,32 @@
...
     protected volatile boolean shutdown;
 
+    protected abstract class ProcessingTask implements Runnable {
+        protected StringBuilder builder = new StringBuilder();
+        protected CharBuffer buffer = CharBuffer.wrap(builder);
+        protected final CharsetEncoder utf8Encoder = MESSAGE_CHARSET.newEncoder()
+                .onMalformedInput(CodingErrorAction.REPLACE)
+                .onUnmappableCharacter(CodingErrorAction.REPLACE);
+
+        public abstract void run();
```
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDProcessor.java]: Indeed, this should just be the common implementation we have in the subclasses.

File: src/main/java/com/timgroup/statsd/StatsDProcessor.java [hunk 5]
```diff
@@ -14,12 +23,15 @@
...
     protected static final Charset MESSAGE_CHARSET = Charset.forName("UTF-8");
+
     protected static final String MESSAGE_TOO_LONG = "Message longer than size of sendBuffer";
     protected static final int WAIT_SLEEP_MS = 10;  // 10 ms would be a 100HZ slice
 
     protected final StatsDClientErrorHandler handler;
 
     protected final BufferPool bufferPool;
+    protected final List<StringBuilder> builders; // StringBuilders for processing, 1 per worker
+    protected final List<CharBuffer> charBuffers; // CharBuffers for processing, 1 per worker
```
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDProcessor.java]: Yes.

File: src/main/java/com/timgroup/statsd/StatsDProcessor.java [hunk 6]
```diff
@@ -39,9 +77,20 @@
...
         this.endSignal = new CountDownLatch(workers);
+
+        this.builders = new ArrayList<>(workers);
+        this.charBuffers = new ArrayList<>(workers);
+        for (int i = 0 ; i < workers ; i++) {
+            StringBuilder builder = new StringBuilder();
+            CharBuffer buffer = CharBuffer.wrap(builder);
+            builders.add(builder);
+            charBuffers.add(buffer);
+        }
```
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDProcessor.java]: And Yes.

File: src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java [hunk 7]
```diff
@@ -1,25 +1,107 @@
...
+
+                        sendBuffer.mark();
+                        if (sendBuffer.position() > 0) {
+                            sendBuffer.put((byte) '\n');
+                        }
+
+                        try {
+                            writeBuilderToSendBuffer(sendBuffer);
+                        } catch (BufferOverflowException boe) {
+                            outboundQueue.put(sendBuffer);
```
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java]: Ummm, not sure. It breaks the tests too, I'll dig a little.

File: src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java [hunk 8]
```diff
@@ -1,25 +1,107 @@
...
+                return;
+            }
+
+            while (!(messages.isEmpty() && shutdown)) {
+
+                try {
+
+                    if (Thread.interrupted()) {
+                        return;
+                    }
```
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java]: cleared it out.

File: src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java]: It has no effect, tried a bunch of tests same packet count, same message count. Resetting the bottom one will actually break things. So I believe we don't need the resets at all.

**Pull request merged**